libbfddp/bfddp_extra.c

```diff
@@ -0,0 +1,858 @@
+	bs->bs_rstate = state;
+	bs->bs_rdiag = bcp->version_diag & 0x1F;
+	bs->bs_rid = ntohl(bcp->local_id);
+
+	/* Detect timers change: */
+	if (ntohl(bcp->desired_tx) != bs->bs_rtx)
+		timers_changed = true;
+	else if (ntohl(bcp->required_rx) != bs->bs_rrx)
+		timers_changed = true;
+	else if (ntohl(bcp->required_echo_rx) != bs->bs_rerx)
+		timers_changed = true;
+	else if (bcp->detection_multiplier != bs->bs_rdmultiplier)
+		timers_changed = true;
+
+	bs->bs_rtx = ntohl(bcp->desired_tx);
+	bs->bs_rrx = ntohl(bcp->required_rx);
+	bs->bs_rerx = ntohl(bcp->required_echo_rx);
+	bs->bs_rdmultiplier = bcp->detection_multiplier;
+	bs->bs_rcbit = !!(bcp->state_bits & STATE_CPI_BIT);
+	bs->bs_rdemand = !!(bcp->state_bits & STATE_DEMAND_BIT);
+
+	/* Skip the rest if this session is shutdown. */
+	if (bs->bs_admin_shutdown)
+		return;
+
+	/*
+	 * RFC 5880 Section 6.8.6. Reception of BFD Control Packets:
+	 *
+	 * > If a Poll Sequence is being transmitted by the local system
+	 * > and the Final (F) bit in the received packet is set, the
+	 * > Poll Sequence MUST be terminated.
+	 */
+	if (bs->bs_poll && (bcp->state_bits & STATE_FINAL_BIT)) {
+		bs->bs_poll = false;
+		bs->bs_final = false;
+
+		/* Negotiation ended, apply new intervals. */
+		bs->bs_cur_dmultiplier = bs->bs_dmultiplier;
+		bs->bs_cur_tx = bs->bs_tx;
+		bs->bs_cur_rx = bs->bs_rx;
+		bs->bs_cur_erx = bs->bs_erx;
+
+		/* Tell control plane about timers change. */
+		bfddp_send_session_state_change(bs);
+	} else {
+		if (timers_changed) {
+			/*
+			 * Notify control plane about new timers if this is not
+			 * the final event.
+			 */
+			bfddp_send_session_state_change(bs);
+
+			/*
+			 * Update transmission timer to avoid session
+			 * disruption.
+			 *
+			 * NOTE:
+			 * Receive timer update will be called later this
+			 * function.
+			 */
+			bfddp_callbacks.bc_tx_control_update(bs, arg);
+		}
+	}
+
+	/*
+	 * RFC 5880 Section 6.8.6. Reception of BFD Control Packets:
+	 *
+	 * > If the Poll (P) bit is set, send a BFD Control packet to the
+	 * > remote system with the Poll (P) bit clear, and the Final (F)
+	 * > bit set (see section 6.8.7).
+	 */
+	if (bcp->state_bits & STATE_POLL_BIT) {
+		bs->bs_poll = false;
+		bs->bs_final = true;
+
+		/* Speed up session convergence. */
+		bfddp_send_control_packet(bs, arg);
+
+		bs->bs_final = false;
+	}
+
+	bfddp_session_state_machine(bs, arg, bs->bs_rstate);
+
+	/* We received the peer packet, update the expiration timer. */
+	bfddp_callbacks.bc_rx_control_update(bs, arg);
+
+	/* Handle echo timer not implemented. */
+}
+
+/*
+ * FRR BFD communication functions.
+ */
+size_t
+bfddp_send_echo_request(struct bfddp_ctx *bctx)
+{
+	struct bfddp_message msg = {};
+	struct timeval tv;
+
+	/* Prepare header. */
+	msg.header.version = BFD_DP_VERSION;
+	msg.header.length = htons(sizeof(struct bfddp_message_header)
+				  + sizeof(struct bfddp_echo));
+	msg.header.type = htons(ECHO_REQUEST);
+
+	/* Payload data. */
+	gettimeofday(&tv, NULL);
```

Comment:
So, might be important... might not.

`gettimeofday()` is affected by discontinous jumps in the clock wall (e.g. resync of NTP, manual changes on `date`).

I am not sure what tis this time used for, but if it's used for internal callbacks you might want to have a look into `CLOCK_MONOTONIC`, not affected by jumps and significantly cheaper. Take a look into `CLOCK_MONOTONIC_COARSE` too, but this has more portability issues.

Reply:
It is not used for critical things, only for measuring ping latency/processing delay.

We can't use MONOTONIC clock here because each application has its "own" MONOTONIC clock. It gets a bit more complicated when the data plane is running on a different OS/machine than the FRR BFD too.

Reply:
You mean it's shared across different processes, potentially in different machines?

Reply:
Yes.

Here is the FRR BFD code that answers:
https://github.com/opensourcerouting/frr/blob/bfd-distributed-v3/bfdd/dplane.c#L463

Reply:
ACK